pkg/ruler/manager.go

```diff
@@ -201,19 +208,26 @@ func (r *DefaultMultiTenantManager) syncRulesToManager(ctx context.Context, user
 		return
 	}
 
-	if !existing || update {
+	if !existing || rulesUpdated || externalLabelsUpdated {
 		level.Debug(r.logger).Log("msg", "updating rules", "user", user)
 		r.configUpdatesTotal.WithLabelValues(user).Inc()
-		if update && existing {
+		if rulesUpdated && existing {
```

Comment:
I think this needs to be `if (rulesUpdated || externalLabelsUpdated) && existing`. Rule cache exists here so serve the List Rules API requests while Prometheus rule manager is updated. In some cases where the rules are long running, it can take a bit for Prometheus rule manager to stop the rules and update it while holding the lock. If external labels are updated and if that triggers the behavior mentioned above, I still think the cache needs to be updated

Reply:
I might not understand this correctly, but from what I see, external labels only apply after the evaluation and sending the results (alerts only) to external systems. It does not seem to affect the cache.

Reply:
If `externalLabelsUpdated` is true, `manager.Update(r.cfg.EvaluationInterval, files, externalLabels, r.cfg.ExternalURL.String(), ruleGroupIterationFunc)` will be called. This can take a while in cases where the rule groups are time consuming and the Update holds the lock while it's waiting for rule groups to be stopped and updated. During this time, the ListRules calls will have to wait for that lock to be released. The cache was introduced so that ListRules can keep serving rules that were evaluating just prior to the update

Reply:
Oh.. there is a `deleteRuleCache` after the `manager.Update`. I thought the cache always existed.